### PR TITLE
Add boolean attribute main_image_from_frontmatter to indicate if cove…

### PR DESCRIPTION
**Pull request by vishaldeepak**
## What type of PR is this? (check all applicable)

- [x] Feature
- [x] Bug Fix

## Description
Added a new boolean field to articles table. This will enable us to keep track if cover_image/main_image was set from front matter. If so , then if next time we do not receive the field cover_image from front matter, we can clear out the cover image.

## Related Tickets & Documents
Directly related to #15233 

## QA Instructions, Screenshots, Recordings
Altered slightly from the issue reproduction steps
1.  Click on the "Create Post" button
2. Fill in the Jekyll Front Matter (JFM) information for the post (title, published: true, some description, some tag, cover_image). Make sure to set cover_image
3. Save the post (it will be published because JFM published is true)
4. You'll be taken to the published article, now click on the "Edit" button
5. Remove the cover_image line from JFM
6. Save the post (it will be automatically republished because published: true)
7.  Cover image is now removed, earlier  it was not

## Added/updated tests?

- [x] Yes

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?
Migrate the database

### CODE REVIEW

File: spec/requests/articles/articles_update_spec.rb [hunk 1]
```diff
@@ -167,4 +167,43 @@
     expect(response).to redirect_to "#{article.path}/edit"
     expect(article.reload.video_thumbnail_url).to include "https://i.imgur.com/HPiu7N4.jpg"
   end
+
+  context "when main_image_from_frontmatter is false" do
```
Comment (jeremyf) [inline spec/requests/articles/articles_update_spec.rb:171]: These are great specs, I'm wondering if it would be better to put these as model tests.  The request specs spin up a lot of context to run, as compared to model specs.

Thoughts?

Comment (vishaldeepak) [reply spec/requests/articles/articles_update_spec.rb:171]: for sure they can be moved. Do you think its worth keeping the once which directly resolves the issue. But to be fair that can be moved too.

Comment (jeremyf) [reply spec/requests/articles/articles_update_spec.rb:171]: My personal preference is to have that be a model test.  Those request tests are slow, and if we already cover the issue with a model test, I'd like to avoid adding a slow test to the suite.

Comment (vishaldeepak) [reply spec/requests/articles/articles_update_spec.rb]: @jeremyf done as required

File: spec/models/article_spec.rb [hunk 2]
```diff
@@ -708,6 +708,41 @@ def build_and_validate_article(*args)
...
+        article.update! body_markdown: "---\ntitle: hey\npublished: false\n---\nYo ho ho#{rand(100)}"
+        expect(article.reload.main_image).not_to be_nil
+      end
+
+      it "does remove main image if cover image is passed empty in markdown" do
+        expect(article.main_image).not_to be_nil
+        article.update! body_markdown: "---\ntitle: hey\npublished: false\ncover_image: \n---\nYo ho ho#{rand(100)}"
+        expect(article.reload.main_image).to be_nil
+      end
+    end
```
Comment (jgaskins) [inline spec/models/article_spec.rb:733]: These specs don't reference `main_image_from_frontmatter` at all, so they're relying on containing RSpec contexts. It seems we're assuming them to be `false` here because that's the default value if not provided and the containing contexts currently don't set them, but it's not guaranteed inside the specs in this context. If `article` changes elsewhere in this file to set `main_image_from_frontmatter` (either explicitly or implicitly), these specs could fail for unrelated reasons.

Could you explicitly set this value to `false` inside this context?

File: app/models/article.rb [hunk 3]
```diff
@@ -634,16 +634,16 @@ def evaluate_front_matter(front_matter)
...
-  def determine_image(front_matter)
-    # In order to clear out the cover_image, we check for the key in the front_matter.
-    # If the key exists, we use the value from it (a url or `nil`).
-    # Otherwise, we fall back to the main_image on the article.
-    has_cover_image = front_matter.include?("cover_image")
-
-    if has_cover_image && (front_matter["cover_image"].present? || main_image)
-      front_matter["cover_image"]
+  def set_image(front_matter)
+    if front_matter.include?("cover_image")
```
Comment (jgaskins) [inline app/models/article.rb:638]: I checked whether `cover_image` appearing anywhere else in the front matter (for example, `description: Something something cover_image`) would inadvertently set this and it worked as intended. 👍🏼

**Review by jgaskins — COMMENTED** [review]
Small change requested, but nonblocking

**Comment by jeremyf** [comment]
@vishaldeepak This looks great, and I think is acceptable as is.

However there are a few considerations as I look through this with a "final" edit mindset:

1. Renaming the method name
2. Flipping the logic
3. Existing data considerations

## Renaming the Method Name

Given that we allow multiple images, though only one "main_image", I would like to see the method name `set_main_image`.  This helps clarify that it's not for setting "any old image".

Also, replacing `elsif main_image_from_frontmatter` with `elsif main_image_from_frontmatter?`.  Endinig the method with `?` is more Ruby idiomatic (and something I'm trying to encourage in our Ruby methods).

## Flipping the Logic

With your proposed change we have the following logic:

```ruby
def set_image(front_matter)
  if front_matter.include?("cover_image")
    self.main_image = front_matter["cover_image"]
    self.main_image_from_frontmatter = true
  elsif main_image_from_frontmatter
    # If we do not receive an image from frontmatter but image was earlier set from frontmatter then clear image
    self.main_image = nil
    self.main_image_from_frontmatter = true
  else
    self.main_image_from_frontmatter = false
  end
end
```

I'm thinking about flipping the logic to drive on the indicator instead of first interrogating the front_matter:

```ruby
# @param front_matter [Hash]
#
# @note There's an assumption, as enforced in the database, that
#       `articles.main_image_from_frontmatter` defaults to `false`.
def set_image(front_matter)
  # At one point, we have set the image_image based on the front matter.
  # Forever will that now dictate the behavior.
  if main_image_from_front_matter?
    self.main_image = front_matter["cover_image"]
  elsif front_matter.key?("cover_image")
	# They've chosen the set cover image in the front matter, so we'll
	# proceed with that assumption.
    self.main_image = front_matter["cover_image"]
    self.main_image_from_frontmatter = true
  end
end
```

How does my flipping of the logic look?  Aside from the comments, I think that it reads easier (but that's my author bias).

## Existing Data Considerations

Do we need to write a [data update script](https://developers.forem.com/backend/data-update-scripts)?

In other words, what kinds of disruptions might we introduce in adding this flag?

We have two scenarios to consider:

1.  The user goes to their article and updates it.
2.  Administrative tasks update an article (and evaluates the markdown and front-matter).

In the user updates the article scenario, I think the visual feedback would be adequate for helping them navigate any possible "disruption."

In the administrative task update scenario, my hesitation is that we're introducing an indicator that's a bit irreversible (e.g. there's no real logic path out from the `main_image_from_frontmatter == true`).  I don't think that's an issue.

**Comment by vishaldeepak** [comment]
@jeremyf 
1. Renaming the method name - Sounds good, will change it 
2. Flipping the logic - Your code looks much cleaner and readable. But I have  one scenario that may not work
 - User creates article and sets main_image. Flag main_image_from_frontmatter -> true
 - User updates articles to remove main_image. Flag main_image_from_frontmatter still -> true
 - Image is now set via API(admin, directly set main_image param) on the article. Flag main_image_from_frontmatter still -> true : **Should be false**
 Dont know if this case is possible.
3. Existing data considerations - I think we should run a data update task to check if existing articles have a valid `cover_image` attribute and then set `main_image_from_frontmatter` flag. This should keep the data consistent for new and existing articles

File: app/models/article.rb (hunk 3 above)
Comment (jeremyf) [reply app/models/article.rb:638]: @jgaskins I think there's some antics with `.include?`.  The `front_matter` is a Hash, and `.include?` just checks the keys.  My recommended refactor is to move from `.include?` to `.key?`

**Comment by jeremyf** [comment]
> Flipping the logic - Your code looks much cleaner and readable. But I have one scenario that may not work
>
> * User creates article and sets main_image. Flag main_image_from_frontmatter -> true
> * User updates articles to remove main_image. Flag main_image_from_frontmatter still -> true
> * Image is now set via API(admin, directly set main_image param) on the article. Flag main_image_from_frontmatter still -> true : Should be false
> 
> Dont know if this case is possible.

I'm jumping between several contexts outside of this issue, so I want to make sure I'm understanding.  Does my code which looks cleaner not resolve your above scenario?  I believe that to be the case.

Importantly, does your code resolve the above scenario?

My gut in this is that an indicator, such as `main_image_from_frontmatter`, will be hard to manage this subtle state-machine.

Fundamentally, the code (prior to the change) says use `front_matter` to overwrite `main_image`.  The `front_matter` looks to clobber most of the `main_image`.

This is a scenario where we want an intermediary Ruby form object to translate the posted front matter into attributes for the Article.  That is to say we _ideally_ don't want the Article to make decisions about should the `front_matter` overwrite `main_image` or not.

I think we can approximate a Ruby form object by leveraging something from the [ActiveRecord::AttributeMethods::Dirty](https://api.rubyonrails.org/classes/ActiveRecord/AttributeMethods/Dirty.html) module.

**Comment by vishaldeepak** [comment]
@jeremyf Apologies I didnt looks at your comment
```
  # At one point, we have set the image_image based on the front matter.
  # Forever will that now dictate the behavior.
```
My case was a possibility, if main_image was set from other source(other than front_matter) even after it was set once from front_matter. So we should be good with that. 

Il update the PR. What do you think of the point 3? Want me to write a task for that or is it something taken care of by the core team?

**Comment by jeremyf** [comment]
@vishaldeepak if you would be able to write the data script that would be great.  Barring that, laying out the logic.

The data scripts are a bit particular in that we have to be mindful of query timeouts.  But, if you get the logic around and the general concept I'm sure I can get some core team folks to help carry this forward.

**Comment by vishaldeepak** [comment]
@jeremyf  changes made , added data script as well

**Comment by jeremyf** [comment]
@vishaldeepak sorry about the delay in getting back to you.  Things look good.

We have a schema conflict.  And I suspect the data update script will need to use Async workers.  My understanding is that we could easily hit a time out in processing articles.

I'm happy to carry it forward from here.  Please let me know your preference.

**Comment by vishaldeepak** [comment]
@jeremyf Yes , you can take it forward , thank you

**Comment by citizen428** [comment]
> And I suspect the data update script will need to use Async workers. 

During deploys we enqueue a worker (`DataUpdateWorker`) to run data updates (see `lib/tasks/data_updates.rake`). So unless you want more granular workers (e.g. one per update for easier restarts/monitoring) you may not have to do anything in this regard.

File: app/models/article.rb [hunk 4]
```diff
@@ -642,16 +642,14 @@ def evaluate_front_matter(front_matter)
...
-    # If the key exists, we use the value from it (a url or `nil`).
-    # Otherwise, we fall back to the main_image on the article.
-    has_cover_image = front_matter.include?("cover_image")
-
-    if has_cover_image && (front_matter["cover_image"].present? || main_image)
-      front_matter["cover_image"]
-    else
-      main_image
+  def set_main_image(front_matter)
+    # At one point, we have set the main)image based on the front matter. Forever will that now dictate the behavior.
```
Comment (jeremyf) [inline app/models/article.rb:646]:
```suggestion
    # At one point, we have set the main_image based on the front matter. Forever will that now dictate the behavior.
```

**Review by jeremyf — APPROVED** [review]

**Review by citizen428 — APPROVED** [review]
